src/ApparentHorizons/YlmSpherepack.cpp

```diff
@@ -689,106 +689,103 @@ void YlmSpherepack::interpolate(
 }
 /// \endcond
 
-template <typename T>
+template <typename T, typename R>
 void YlmSpherepack::interpolate_from_coefs(
-    const gsl::not_null<std::vector<double>*> result, const T& spectral_coefs,
-    const InterpolationInfo& interpolation_info, const size_t spectral_stride,
-    const size_t spectral_offset) const noexcept {
+    const gsl::not_null<T*> result, const R& spectral_coefs,
+    const InterpolationInfo<T>& interpolation_info,
+    const size_t spectral_stride, const size_t spectral_offset) const noexcept {
   const auto& alpha = storage_.work_interp_alpha;
   const auto& beta = storage_.work_interp_beta;
   const auto& index = storage_.work_interp_index;
   // alpha holds alpha(n,m,x)/x, beta holds beta(n+1,m).
   // index holds the index into the coefficient array.
   // All are indexed together.
 
-  ASSERT(result->size() == interpolation_info.size(),
-         "Size mismatch: " << result->size() << ","
-                           << interpolation_info.size());
+  const size_t num_points = get_size(*result);
+  ASSERT(num_points == interpolation_info.size(),
+         "Size mismatch: " << num_points << "," << interpolation_info.size());
 
   const size_t l1 = m_max_ + 1;
 
   // Offsets of 'a' and 'b' in spectral_coefs.
   const size_t a_offset = spectral_offset;
   const size_t b_offset = spectral_offset + (l1 * n_theta_) * spectral_stride;
 
-  std::fill(result->begin(), result->end(), 0.0);
+  const auto& cos_theta = interpolation_info.cos_theta;
 
-  for (size_t i = 0; i < result->size(); ++i) {
-    const auto& intrp_info = interpolation_info[i];
-    const auto& cos_theta = intrp_info.cos_theta;
+  *result = make_with_value<T>(cos_theta, 0.);
 
-    // Clenshaw recurrence for m=0.  Separate because there is no phi
-    // dependence, and there is a factor of 1/2.
-    size_t idx = 0;
-    {
-      double ycn = 0.0;
-      double ycnp1 = 0.0;
-      for (size_t n = n_theta_ - 1; n > 0;
-           --n, ++idx) {  // Loops from n_theta_-1 to 1.
-        double ycnp2 = ycnp1;
-        ycnp1 = ycn;
-        ycn = cos_theta * alpha[idx] * ycnp1 + beta[idx] * ycnp2 +
-              spectral_coefs[a_offset + spectral_stride * index[idx]];
-      }
-      (*result)[i] += 0.5 * intrp_info.pbar_factor[0] *
-                      (beta[idx] * ycnp1 + cos_theta * alpha[idx] * ycn +
-                       spectral_coefs[a_offset + spectral_stride * index[idx]]);
-      ++idx;
+  // Clenshaw recurrence for m=0.  Separate because there is no phi
+  // dependence, and there is a factor of 1/2.
+  size_t idx = 0;
+  {
+    T ycn = make_with_value<T>(cos_theta, 0.);
+    T ycnp1 = make_with_value<T>(cos_theta, 0.);
+    for (size_t n = n_theta_ - 1; n > 0;
+         --n, ++idx) {  // Loops from n_theta_-1 to 1.
+      T ycnp2 = ycnp1;
```

Comment:
again, don't do memory allocations in a for loop, that's around the worst thing you can do for multithreaded systems.